circuits/src/zk_gadgets/poseidon.rs

```diff
@@ -142,451 +127,217 @@ impl PoseidonHashGadget {
-    state: Vec<MpcLinearCombination>,
-    /// The next index in the state to being absorbing inputs at
-    next_index: usize,
-    /// Whether the sponge is in squeezing mode. For simplicity, we disallow
-    /// the case in which a caller wishes to squeeze values and the absorb more.
-    in_squeeze_state: bool,
-    /// A reference to the shared MPC fabric that the computation variables are
-    /// allocated in
-    fabric: MpcFabric,
-}
-
-impl MultiproverPoseidonHashGadget {
-    /// Construct a new hash gadget with the given parameterization
-    pub fn new(params: PoseidonParams, fabric: MpcFabric) -> Self {
-        // Initialize the state as all zeros
-        let state = (0..params.capacity + params.rate)
-            .map(|_| MpcLinearCombination::from_scalar(Scalar::zero(), fabric.clone()))
-            .collect::<Vec<_>>();
-        Self {
-            params,
-            state,
-            next_index: 0,
-            in_squeeze_state: false, // Start in absorb state
-            fabric,
-        }
-    }
-
-    /// Hashes the payload and then constrains the squeezed output to be the
-    /// provided expected output.
-    pub fn hash<L, CS>(
+    /// Apply the external MDS matrix to the state
+    ///
+    /// For t = 3, this is the circulant matrix `circ(2, 1, 1)`
+    ///
+    /// This is equivalent to doubling each element then adding the other two to
+    /// it, or more efficiently: adding the sum of the elements to each
+    /// individual element. This efficient structure is borrowed from:
+    ///     https://github.com/HorizenLabs/poseidon2/blob/main/plain_implementations/src/poseidon2/poseidon2.rs#L129-L137
+    fn external_mds<C: ConstraintSystem<ScalarField>>(
         &mut self,
-        hash_input: &[L],
-        expected_output: &L,
-        cs: &mut CS,
-    ) -> Result<(), ProverError>
-    where
-        L: MpcLinearCombinationLike,
-        CS: MpcRandomizableConstraintSystem,
-    {
-        self.batch_absorb(hash_input, cs)?;
-        self.constrained_squeeze(expected_output.clone(), cs)
-    }
-
-    /// Absorb an input into the hasher state
-    pub fn absorb<L, CS>(&mut self, a: L, cs: &mut CS) -> Result<(), ProverError>
-    where
-        L: MpcLinearCombinationLike,
-        CS: MpcRandomizableConstraintSystem,
-    {
-        assert!(
-            !self.in_squeeze_state,
-            "Cannot absorb from a sponge that has already been squeezed"
-        );
-
-        // Permute the digest state if we have filled up the rate sized buffer
-        if self.next_index == self.params.rate {
-            self.permute(cs)?;
-            self.next_index = 0;
+        cs: &mut C,
+    ) -> Result<(), PlonkError> {
+        let coeffs = [ScalarField::one(); GATE_WIDTH];
+        let in_wires = self.state.clone();
+        for state_elem in self.state.iter_mut() {
+            let lc_wires = [*state_elem, in_wires[0], in_wires[1], in_wires[2]];
+            *state_elem = cs.lc(&lc_wires, &coeffs)?;
```

Comment:
would it be more appropriate to use `cs.sum` here?

Reply:
No, `cs.sum` will break up the input into groups of 3 inputs at a time, so this would require two gates instead of one.